Tracking critic dwb

silanus23 opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   |#5037|
| Primary OS tested on | Ubuntu |
| Robotic platform tested on |nav2_bringup |
| Does this PR contain AI generated software? | util tests|
| Was this PR description generated by AI software? | nope definitely not |

---

## Description of contribution in a few bullet points

<!--
* I added this neat new feature
* Also fixed a typo in a parameter name in nav2_costmap_2d
-->

## Description of documentation updates required from your changes
Critic additons and their parameter additions
<!--
* Added new parameter, so need to add that to default configs and documentation page
* I added some capabilities, need to document them
-->

## Description of how this change was tested
Added util tests. I have succeded to create cases where turtlebot has to create a new path cause it's so out of bounds(done this on dwb).
<!--
* I wrote unit tests that cover 90%+ of changes and extensively tested on my physical robot platform in production for 1 week
* I wrote unit tests and tested in simulation for 10 minutes
* Performed linting validation using pre-commit run --all or colcon test
-->

---

## Future work that may be required in bullet points
These can be added to default yaml
<!--
* I think there might be some optimizations to be made from STL vector
* I see a lot of redundancy in this package, we might want to add a function `bool XYZ()` to reduce clutter
* I tested on a differential drive robot, but there might be issues turning near corners on an omnidirectional platform
-->
###Note: 
I had to create a new branch cause the original branch was sub branch of my first PR. I had to carry it via copy paste couldn't find a better way. 

#### For Maintainers: <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists
- [ ] Should this be backported to current distributions? If so, tag with `backport-*`.

silanus23 committed:
DWB critics addition

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
mppi critic addition

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
Taking back later additions creating problem

Signed-off-by: silanus23 <berkantali23@outlook.com>

SteveMacenski commented on nav2_dwb_controller/dwb_critics/include/dwb_critics/path_hug.hpp:
```diff
@@ -0,0 +1,51 @@
+protected:
+  bool zero_scale_{false};
+  nav_msgs::msg::Path global_path_;
+  size_t start_index_{0};
```
Nor this

SteveMacenski reviewed (commented):
Otherwise, I can't comment too too much on the formulation without you/I doing some some simulation testing to see how well functionally it works.

silanus23 committed:
addressing reviews and adding bideractional window

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
fixing absolute value problem

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
update with mppi foundings: trajectory and path min distancing, add heading error, have backward search and implement own deviation calculator to achive more control. Update tests accordingly.

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 commented:
Hey I have implemented this version from mppi's improvements. Since this is an alternative and can work without path align I have added a basic heading critic too. Right now it seems to be working and it's performance is better than path align. Unlike mppi I didn't come across any velocity issue.
While I draw a very optimistic outcome I must say I didn't push this code much. I am trying this on tb3 which since it's more harmonized with dwb than tb4. I think this is a good base we can develop on. 
I will try to push videos before going offline but not sure if I can
Ohh lastly I know you want a diffirent name. I have 4 alternatives at the top my head. Path adhesive, tracking error, bounded tracking error, path tracking. I can change to one of these or to your alternative if you want, both mppi and this.
